README.md
# Declarative `setuptools` Config Requirements Files Support

This projects adds the ability for projects using setuptools declarative configuration
to specify requirements using requirements files.

### ⚠ **There's a reason why this isn't, at least yet, supported by default. Please [read why](https://github.com/pypa/setuptools/issues/1951).**

Anyway, if you know what you're doing, then this library solves the missing feature
of defining requirements using requirements files.

## `setup.cfg`
Your `setup.cfg` should include a section named `requirements-files`, like:

```ini
[requirements-files]
setup_requires = requirements/setup.txt
install_requires = requirements/base.txt
tests_require = requirements/tests.txt
extras_require =
  docs = requirements/docs.txt
  cli = requirements/cli.txt
```

### ⚠ ATTENTION

#### The requirements files **MUST** be included in the wheel file aswell as the source tarball

For the example shown above, in ``setup.cfg``, add something like:

```ini
[options.data_files]
. = requirements/*.txt
```

Or something like the folowing on your ``MANIFEST.in``:

```
include requirements/*.txt
```

Or, if you use [setuptools-scm](https://pypi.org/project/setuptools-scm), the requirements files
need to be committed to the SCM repo.


## `pyproject.toml`
Your `pyproject.toml` should also include `setuptools-declarative-requirements`:
```toml
[build-system]
requires = ["setuptools>=50.3.2", "wheel", "setuptools-declarative-requirements"]
build-backend = "setuptools.build_meta"
```

## `setup.py`
Some projects still use a `setup.py` shim, similar to:
```python
#!/usr/bin/env python
import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

If this is your case, your `setup.cfg` needs an extra bit of tweak. Make sure you have
``setuptools-declarative-requirements`` in your `setup_requires`:
```ini
[options]
setup_requires =
  setuptools>=50.3.2
  setuptools-declarative-requirements
```

## Do Note That
### ⚠ This project makes no attempt to validate your requirements files.
**The only thing it does is include every non empty line from your requirements files which does
not start with `#`, `-r` or `--`.**


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.0](https://pypi.org/project/setuptools-declarative-requirements/1.3.0/) - 2022-08-02
### Changed
- Ignore lines that start with `-i`. Fixes #6

## [1.2.0](https://pypi.org/project/setuptools-declarative-requirements/1.2.0/) - 2020-12-19
### Changed
- Exclude the tests package from being installed

## [1.1.0](https://pypi.org/project/setuptools-declarative-requirements/1.1.0/) - 2020-10-29
### Changed
- Drop support down to Py3.5

## [1.0.0](https://pypi.org/project/setuptools-declarative-requirements/1.0.0/) - 2020-10-29
### Changed
- Improve the README config details

## [1.0.0rc3](https://pypi.org/project/setuptools-declarative-requirements/1.0.0rc3/) - 2020-10-29
### Changed
- The project now uses `setup.cfg` instead of `pyproject.toml`
- Fixed the project urls

## [1.0.0rc2](https://pypi.org/project/setuptools-declarative-requirements/1.0.0rc2/) - 2020-10-28
### Added
- Fixed the README.md example
- Added a CHANGELOG.md

## [1.0.0rc1](https://pypi.org/project/setuptools-declarative-requirements/1.0.0rc1/) - 2020-10-28
### Added
- First working version of the project


tests/conftest.py
import logging
import os
import pathlib
import pprint
import random
import shutil
import string
import subprocess
import sys
import textwrap
from contextlib import contextmanager

import attr
import pytest
from pytestshellutils.shell import Subprocess
from pytestshellutils.utils import ports

from declarative_requirements import __version__

CODE_ROOT = pathlib.Path(__file__).resolve().parent.parent

log = logging.getLogger(__name__)


def random_string(prefix, size=6, uppercase=False, lowercase=True, digits=True):
    """
    Generates a random string.

    Args:
        prefix(str): The prefix for the random string
        size(int): The size of the random string
        uppercase(bool): If true, include upper-cased ascii chars in choice sample
        lowercase(bool): If true, include lower-cased ascii chars in choice sample
        digits(bool): If true, include digits in choice sample
    Returns:
        str: The random string
    """
    if not any([uppercase, lowercase, digits]):  # pragma: no cover
        raise RuntimeError(
            "At least one of 'uppercase', 'lowercase' or 'digits' needs to be true"
        )
    choices = []
    if uppercase:
        choices.extend(string.ascii_uppercase)
    if lowercase:
        choices.extend(string.ascii_lowercase)
    if digits:
        choices.extend(string.digits)

    return prefix + "".join(random.choice(choices) for _ in range(size))


@attr.s
class VirtualEnv:

    env = attr.ib()
    venv_dir = attr.ib()
    venv_python = attr.ib(init=False)

    @venv_dir.validator
    def _venv_dir(self, _, value):
        return value / random_string("venv-")

    @venv_python.default
    def _venv_python(self):
        if sys.platform.startswith("win"):
            return self.venv_dir / "Scripts" / "python.exe"
        else:
            return self.venv_dir / "bin" / "python"

    @property
    def venv_bin_dir(self):
        return self.venv_python.parent

    def __enter__(self):
        try:
            self._create_virtualenv()
        except subprocess.CalledProcessError:
            raise AssertionError("Failed to create virtualenv")
        return self

    def __exit__(self, *args):
        shutil.rmtree(str(self.venv_dir))

    def install(self, *args, **kwargs):
        return self.run(str(self.venv_python), "-m", "pip", "install", *args, **kwargs)

    def run(self, *cmdline, cwd=None, env=None):
        if cwd is None:
            cwd = str(self.venv_dir)
        if env is None:
            env = os.environ.copy()
        env.update(self.env)
        return Subprocess(cwd=cwd, environ=env).run(*cmdline)

    def get_installed_packages(self):
        ret = self.run(str(self.venv_python), "-m", "pip", "list", "--format", "json")
        pkgs = {}
        for entry in ret.data:
            pkgs[entry["name"].lower()] = entry
        return pkgs

    def _get_real_python(self):
        """
        The reason why the virtualenv creation is proxied by this function is mostly
        because under windows, we can't seem to properly create a virtualenv off of
        another virtualenv(we can on linux) and also because, we really don't want to
        test virtualenv creation off of another virtualenv, we want a virtualenv created
        from the original python.
        Also, on windows, we must also point to the virtualenv binary outside the
        existing virtualenv because it will fail otherwise
        """
        try:
            if sys.platform.startswith("win"):
                return os.path.join(sys.real_prefix, os.path.basename(sys.executable))
            else:
                python_binary_names = [
                    "python{}.{}".format(*sys.version_info),
                    "python{}".format(*sys.version_info),
                    "python",
                ]
                for binary_name in python_binary_names:
                    python = os.path.join(sys.real_prefix, "bin", binary_name)
                    if os.path.exists(python):
                        break
                else:
                    bin_dir = os.path.join(sys.real_prefix, "bin")
                    raise AssertionError(
                        "Couldn't find a python binary name under '{bin_dir}' "
                        "matching: {python_binary_names}".format(
                            bin_dir=bin_dir, python_binary_names=python_binary_names
                        )
                    )
                return python
        except AttributeError:
            return sys.executable

    def _create_virtualenv(self):
        virtualenv = shutil.which("virtualenv")
        if virtualenv is None:
            pytest.fail("The 'virtualenv' binary was not found")
        self.run(virtualenv, "--python", self._get_real_python(), str(self.venv_dir))


@attr.s(frozen=True)
class Project:

    cwd = attr.ib()
    env = attr.ib()

    def run(self, *cmdline, env=None):
        if env is None:
            env = os.environ.copy()
        env.update(self.env)
        return Subprocess(cwd=self.cwd, environ=env).run(*cmdline)

    def sys_exec_run(self, *cmdline):
        return self.run(sys.executable, *cmdline)

    def write_file(self, path, contents=None, strip_first_newline=True):
        file_path = self.cwd.joinpath(path)
        file_directory = file_path.parent
        if not file_directory.is_dir():
            file_directory.mkdir(parents=True)
        file_path.touch(exist_ok=True)

        if contents is not None:
            if contents:
                if contents.startswith("\n") and strip_first_newline:
                    contents = contents[1:]
                file_contents = textwrap.dedent(contents)
            else:
                file_contents = contents

            file_path.write_text(file_contents)

    def write_setup(
        self,
        setup_requires=None,
        install_requires=None,
        extras_require=None,
        tests_require=None,
    ):
        if install_requires is None:
            install_requires = ""
        else:
            install_requires = "\n        install_requires = {}".format(
                install_requires
            )
        if setup_requires is None:
            setup_requires = ""
        else:
            setup_requires = "\n        setup_requires = {}".format(setup_requires)
        if tests_require is None:
            tests_require = ""
        else:
            tests_require = "\n        tests_require = {}".format(tests_require)
        if extras_require is None:
            extras_require = ""
        else:
            _extras_require = ""
            for key, value in extras_require.items():
                _extras_require += """          {} = {}\n""".format(key, value)
            if _extras_require:
                _extras_require = """\n        extras_require =\n{}\n""".format(
                    _extras_require
                )
            extras_require = _extras_require
        setup_cfg = """
        [metadata]
        name = test-project
        description = Test project for file support for setuptools declarative setup.cfg
        author = Pedro Algarvio
        author_email = pedro@algarvio.me
        url = https://github.com/s0undt3ch/setuptools-declarative-requirements
        project_urls =
            Source=https://gitlab.com/s0undt3ch/setuptools-declarative-requirements
            Tracker=https://gitlab.com/s0undt3ch/setuptools-declarative-requirements/issues
        license = Apache Software License 2.0
        classifiers =
            Programming Language :: Python
            Programming Language :: Cython
            Programming Language :: Python :: 3
            Programming Language :: Python :: 3 :: Only
            Programming Language :: Python :: 3.5
            Programming Language :: Python :: 3.6
            Programming Language :: Python :: 3.7
            Programming Language :: Python :: 3.8
            Programming Language :: Python :: 3.9
            Development Status :: 4 - Beta
            Intended Audience :: Developers
            License :: OSI Approved :: Apache Software License
        platforms = unix, linux, osx, cygwin, win32

        [options]
        zip_safe = False
        include_package_data = True
        packages = find:
        python_requires = >= 3.5
        install_requires =
          setuptools
          setuptools_declarative_requirements

        [requirements-files]{setup_requires}{install_requires}{tests_require}{extras_require}
        """.format(
            setup_requires=setup_requires,
            install_requires=install_requires,
            tests_require=tests_require,
            extras_require=extras_require,
        )
        self.write_file("setup.cfg", contents=setup_cfg)
        setup = """
        import setuptools
        if __name__ == "__main__":
            setuptools.setup()
        """
        self.write_file("setup.py", contents=setup)
        pyproject = """
        [build-system]
        requires = [
            "setuptools>=42",
            "wheel",
            "setuptools_declarative_requirements=={__version__}"
        ]
        build-backend = "setuptools.build_meta"
        """.format(
            __version__=__version__
        )
        self.write_file("pyproject.toml", contents=pyproject)
        manifest = """
        include requirements/*.txt
        """
        self.write_file("MANIFEST.in", contents=manifest)

    def create_pkg_tree(
        self,
        setup_requires=None,
        install_requires=None,
        extras_require=None,
        tests_require=None,
    ):
        self.write_setup(
            setup_requires=setup_requires,
            install_requires=install_requires,
            extras_require=extras_require,
            tests_require=tests_require,
        )
        self.write_file("tpkg/__init__.py")

        # if os.path.exists("/tmp/proj"):
        #     shutil.rmtree("/tmp/proj")
        # shutil.copytree(self.cwd, "/tmp/proj")

    def get_generated_dist(self):
        dist_dir = self.cwd / "dist"
        dist = next(dist_dir.glob("*.*"))
        return dist.relative_to(self.cwd)

    @contextmanager
    def virtualenv(self):
        with VirtualEnv(venv_dir=self.cwd, env=self.env) as venv:
            yield venv


@pytest.fixture(scope="session")
def local_pypi_repo_path(tmp_path_factory):
    yield tmp_path_factory.mktemp("pypi-pkgs")


@pytest.fixture(scope="session")
def build_test_pkgs(tmp_path_factory, local_pypi_repo_path):
    # Build this project's whell
    subprocess.run([sys.executable, "setup.py", "bdist_wheel"], cwd=str(CODE_ROOT))
    for pkg in CODE_ROOT.joinpath("dist").glob("*.*"):
        shutil.copyfile(str(pkg), str(local_pypi_repo_path.joinpath(pkg.name)))

    # Build the test project's wheel's
    setup_py = textwrap.dedent(
        """\
        import setuptools
        if __name__ == "__main__":
            setuptools.setup()
        """
    )
    pyproject_toml = textwrap.dedent(
        """\
        [build-system]
        requires = ["setuptools>=42", "wheel"]
        build-backend = "setuptools.build_meta"
        """
    )
    setup_cfg_tpl = textwrap.dedent(
        """\
        [metadata]
        name = {name}
        version = 1.0.0
        description = Test project for file support for setuptools declarative setup.cfg
        author = Pedro Algarvio
        author_email = pedro@algarvio.me
        license = Apache Software License 2.0

        [options]
        zip_safe = False
        include_package_data = True
        packages = find:
        """
    )
    for name in (
        "setup-requires",
        "install-requires",
        "docs-extras-require",
        "cli-extras-require",
        "tests-require",
    ):
        src = tmp_path_factory.mktemp("{}-pkg".format(name))
        src.joinpath("setup.py").write_text(setup_py)
        src.joinpath("pyproject.toml").write_text(pyproject_toml)
        src.joinpath("setup.cfg").write_text(
            setup_cfg_tpl.format(name="{}-pkg".format(name))
        )
        pkg = src / "{}_pkg".format(name.replace("-", "_"))
        pkg.mkdir()
        pkg.joinpath("__init__.py").touch()
        subprocess.run(
            [sys.executable, "setup.py", "bdist_wheel"], cwd=str(src), check=True
        )
        for pkg in src.joinpath("dist").glob("*.*"):
            shutil.copyfile(str(pkg), str(local_pypi_repo_path.joinpath(pkg.name)))


@pytest.fixture(scope="session")
def pypi_server_port():
    return ports.get_unused_localhost_port()


@pytest.fixture(scope="session")
def environ(pypi_server_port):
    return {
        "PIP_DISABLE_PIP_VERSION_CHECK": "1",
        "PIP_EXTRA_INDEX_URL": "http://localhost:{}".format(pypi_server_port),
    }


@pytest.fixture(scope="session")
def pypi_server(local_pypi_repo_path, build_test_pkgs, pypi_server_port):
    log.debug(
        "Local PyPi Server Directory Contents:\n%s",
        pprint.pformat(list(local_pypi_repo_path.glob("*"))),
    )
    proc = subprocess.Popen(
        ["pypi-server", "-p", str(pypi_server_port), str(local_pypi_repo_path)]
    )
    try:
        yield
    finally:
        with proc:
            proc.terminate()


@pytest.fixture
def project(tmp_path, pypi_server, environ):
    _cwd = tmp_path / "cwd"
    _cwd.mkdir()
    return Project(cwd=_cwd, env=environ)


tests/integration/test_requirements_parse.py
import pytest


@pytest.fixture
def project(project, pypi_server_port):
    project.write_file(
        "requirements/setup.txt",
        contents="""
        -i http://localhost:{}

        setup-requires-pkg
        """.format(
            pypi_server_port
        ),
    )
    project.write_file(
        "requirements/base.txt",
        contents="""
        --index-url=http://localhost:{}

        install-requires-pkg
        """.format(
            pypi_server_port
        ),
    )
    project.write_file("requirements/docs.txt", contents="docs-extras-require-pkg")
    project.write_file("requirements/cli.txt", contents="cli-extras-require-pkg")
    project.write_file("requirements/tests.txt", contents="tests-require-pkg")
    project.create_pkg_tree(
        setup_requires="requirements/setup.txt",
        install_requires="requirements/base.txt",
        tests_require="requirements/tests.txt",
        extras_require={
            "docs": "requirements/docs.txt",
            "cli": "requirements/cli.txt",
        },
    )
    yield project


@pytest.mark.parametrize("build", ["sdist", "bdist_wheel"])
def test_build_package(project, build):
    ret = project.sys_exec_run("setup.py", build)
    assert ret.returncode == 0
    package = project.get_generated_dist()
    with project.virtualenv() as venv:
        installed_packages = venv.get_installed_packages()
        assert "install-requires-pkg" not in installed_packages
        ret = venv.install(str(package))
        assert ret.returncode == 0, ret
        installed_packages = venv.get_installed_packages()
        assert "install-requires-pkg" in installed_packages

        # Now the docs extras
        installed_packages = venv.get_installed_packages()
        assert "docs-extras-require-pkg" not in installed_packages
        ret = venv.install("{}[docs]".format(package))
        assert ret.returncode == 0, ret
        installed_packages = venv.get_installed_packages()
        assert "docs-extras-require-pkg" in installed_packages

        # Now the cli extras
        installed_packages = venv.get_installed_packages()
        assert "cli-extras-require-pkg" not in installed_packages
        ret = venv.install("{}[cli]".format(package))
        installed_packages = venv.get_installed_packages()
        assert ret.returncode == 0, ret
        assert "cli-extras-require-pkg" in installed_packages


declarative_requirements/integration.py
import pathlib
import sys
from functools import wraps

from distutils import log


def _parse_requirements_file(requirements_file):
    if not requirements_file.is_file():
        log.error(
            "The requirements file '%s' does not exist",
            requirements_file,
        )
        sys.exit(1)
    parsed_requirements = []
    for line in requirements_file.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith(("#", "-i", "-r", "--")):
            continue
        parsed_requirements.append(line)
    log.info(
        "Requirements parsed from '%s': %s",
        requirements_file,
        ", ".join("'{}'".format(req) for req in parsed_requirements),
    )
    return parsed_requirements


def load_declarative_requirements_files(func, dist):
    """
    This function will return a wrapper arround ``dist.parse_config_files``
    in order for us to get the declarative requirements files, AFTER, the
    config files have been parsed
    """

    @wraps(func)
    def wrapper(*args, **kwargs):
        # Parse configuration files
        func(*args, **kwargs)
        # Load the declarative requirements files config
        requirements = dist.command_options.get("requirements-files")
        if not requirements:
            log.info("No 'requirements-files' section was found. Nothing to do.")
            return

        supported_config_keys = (
            "setup_requires",
            "install_requires",
            "extras_require",
            "tests_require",
        )

        for cfgname, (cfgfile, requirements_file) in requirements.items():
            if cfgfile != "setup.cfg":
                # We only support configuring requirements files in setup.cfg
                continue

            if cfgname not in supported_config_keys:
                log.error(
                    "The config key '%s' under 'requirements-files' is not "
                    "supported. Allowed config keys: %s",
                    cfgname,
                    ", ".join("'{}'".format(key) for key in supported_config_keys),
                )
                sys.exit(1)

            if cfgname == "extras_require":
                # Extras require supports mappings, let's parse that out
                if dist.extras_require is None:
                    # If dist.extras_require is still None, make it a dictionary
                    dist.extras_require = {}
                extras_require = dist.extras_require
                for line in requirements_file.splitlines():
                    if not line.strip():
                        # Ignore empty lines
                        continue
                    if "=" not in line:
                        log.error(
                            "Don't know how to parse the line '%s' for extras_require "
                            "under 'requirements-files'. Should be "
                            "'<extras_key> = path/to/requirements.txt'",
                            line.strip(),
                        )
                        sys.exit(1)
                    extras_key, extras_requirements_file = (
                        part.strip() for part in line.split("=")
                    )
                    if extras_key not in extras_require:
                        extras_require[extras_key] = []
                    extras_require[extras_key].extend(
                        _parse_requirements_file(pathlib.Path(extras_requirements_file))
                    )
                continue

            # The rest of the allowed config keys don't support mappings
            if getattr(dist, cfgname) is None:
                # If the dist value for the attribute is still None, make it a list
                setattr(dist, cfgname, [])
            getattr(dist, cfgname).extend(
                _parse_requirements_file(pathlib.Path(requirements_file.strip()))
            )

        # Be sure to call dist._finalize_requires again so that the distribution class
        # set's additional attributes
        dist._finalize_requires()

    return wrapper


def patch_parse_config_files(dist):
    """
    This function will only add a wrapper around ``dist.parse_config_files`` so
    that we can have setuptools/distutils do the config file parsing instead of
    us.
    """
    dist.parse_config_files = load_declarative_requirements_files(
        dist.parse_config_files, dist
    )


# Make our parse_requirements function the last to be evaluated
patch_parse_config_files.order = sys.maxsize


noxfile.py
import datetime
import gzip
import os
import pathlib
import shutil
import tarfile
import tempfile

import nox
from nox.command import CommandFailed

# Nox options
#  Reuse existing virtualenvs
nox.options.reuse_existing_virtualenvs = True
#  Don't fail on missing interpreters
nox.options.error_on_missing_interpreters = False

# Python versions to test against
PYTHON_VERSIONS = ("3", "3.5", "3.6", "3.7", "3.8", "3.9")
# Be verbose when runing under a CI context
CI_RUN = (
    os.environ.get("JENKINS_URL")
    or os.environ.get("CI")
    or os.environ.get("DRONE") is not None
)
PIP_INSTALL_SILENT = CI_RUN is False
SKIP_REQUIREMENTS_INSTALL = "SKIP_REQUIREMENTS_INSTALL" in os.environ
EXTRA_REQUIREMENTS_INSTALL = os.environ.get("EXTRA_REQUIREMENTS_INSTALL")

COVERAGE_VERSION_REQUIREMENT = "coverage==5.2"

# Prevent Python from writing bytecode
os.environ["PYTHONDONTWRITEBYTECODE"] = "1"

# Global Path Definitions
REPO_ROOT = pathlib.Path(__file__).resolve().parent
# Change current directory to REPO_ROOT
os.chdir(str(REPO_ROOT))

# Coverage site customize
SITECUSTOMIZE_DIR = str(REPO_ROOT / "tests" / "support" / "coverage")

ARTIFACTS_DIR = REPO_ROOT / "artifacts"
# Make sure the artifacts directory exists
ARTIFACTS_DIR.mkdir(parents=True, exist_ok=True)
RUNTESTS_LOGFILE = ARTIFACTS_DIR / "runtests-{}.log".format(
    datetime.datetime.now().strftime("%Y%m%d%H%M%S.%f")
)
COVERAGE_REPORT_DB = REPO_ROOT / ".coverage"
COVERAGE_REPORT_PROJECT = ARTIFACTS_DIR.relative_to(REPO_ROOT) / "coverage-project.xml"
COVERAGE_REPORT_TESTS = ARTIFACTS_DIR.relative_to(REPO_ROOT) / "coverage-tests.xml"
JUNIT_REPORT = ARTIFACTS_DIR.relative_to(REPO_ROOT) / "junit-report.xml"


def _get_session_python_version_info(session):
    try:
        version_info = session._runner._real_python_version_info
    except AttributeError:
        session_py_version = session.run_always(
            "python",
            "-c" 'import sys; sys.stdout.write("{}.{}.{}".format(*sys.version_info))',
            silent=True,
            log=False,
        )
        version_info = tuple(
            int(part) for part in session_py_version.split(".") if part.isdigit()
        )
        session._runner._real_python_version_info = version_info
    return version_info


def _get_pydir(session):
    version_info = _get_session_python_version_info(session)
    if version_info < (3, 5):
        session.error("Only Python >= 3.5 is supported")
    return "py{}.{}".format(*version_info)


@nox.session(python=PYTHON_VERSIONS)
def tests(session):
    if SKIP_REQUIREMENTS_INSTALL is False:
        # Always have the wheel package installed
        session.install("wheel", silent=PIP_INSTALL_SILENT)
        session.install(COVERAGE_VERSION_REQUIREMENT, silent=PIP_INSTALL_SILENT)

        # Install requirements
        requirements_file = (
            REPO_ROOT / "requirements" / _get_pydir(session) / "tests.txt"
        )
        install_command = [
            "--progress-bar=off",
            "-r",
            str(requirements_file.relative_to(REPO_ROOT)),
        ]
        session.install(*install_command, silent=PIP_INSTALL_SILENT)

        if EXTRA_REQUIREMENTS_INSTALL:
            session.log(
                "Installing the following extra requirements because the "
                "EXTRA_REQUIREMENTS_INSTALL environment variable was set: "
                "EXTRA_REQUIREMENTS_INSTALL='%s'",
                EXTRA_REQUIREMENTS_INSTALL,
            )
            install_command = ["--progress-bar=off"]
            install_command += [
                req.strip() for req in EXTRA_REQUIREMENTS_INSTALL.split()
            ]
            session.install(*install_command, silent=PIP_INSTALL_SILENT)

        session.install("-e", ".", silent=PIP_INSTALL_SILENT)

    session.run("coverage", "erase")

    python_path_env_var = os.environ.get("PYTHONPATH") or None
    if python_path_env_var is None:
        python_path_env_var = str(SITECUSTOMIZE_DIR)
    else:
        python_path_entries = python_path_env_var.split(os.pathsep)
        if SITECUSTOMIZE_DIR in python_path_entries:
            python_path_entries.remove(SITECUSTOMIZE_DIR)
        python_path_entries.insert(0, SITECUSTOMIZE_DIR)
        python_path_env_var = os.pathsep.join(python_path_entries)

    env = {
        # The updated python path so that sitecustomize is importable
        "PYTHONPATH": python_path_env_var,
        # The full path to the .coverage data file. Makes sure we always write
        # them to the same directory
        "COVERAGE_FILE": str(COVERAGE_REPORT_DB),
        # Instruct sub processes to also run under coverage
        "COVERAGE_PROCESS_START": str(COVERAGE_REPORT_DB.with_name(".coveragerc")),
    }

    args = [
        "--rootdir",
        str(REPO_ROOT),
        "--log-file={}".format(RUNTESTS_LOGFILE),
        "--log-file-level=debug",
        "--show-capture=no",
        "--junitxml={}".format(JUNIT_REPORT),
        "--showlocals",
        "-ra",
        "-s",
    ]
    if session._runner.global_config.forcecolor:
        args.append("--color=yes")
    if not session.posargs:
        args.append("tests/")
    else:
        for arg in session.posargs:
            if arg.startswith("--color") and args[0].startswith("--color"):
                args.pop(0)
            args.append(arg)
        for arg in session.posargs:
            if arg.startswith("-"):
                continue
            if arg.startswith("tests{}".format(os.sep)):
                break
            try:
                pathlib.Path(arg).resolve().relative_to(REPO_ROOT / "tests")
                break
            except ValueError:
                continue
        else:
            args.append("tests/")
    try:
        session.run("coverage", "run", "-m", "pytest", *args, env=env)
    finally:
        # Always combine and generate the XML coverage report
        try:
            session.run("coverage", "combine")
        except CommandFailed:
            # Sometimes some of the coverage files are corrupt which would
            # trigger a CommandFailed exception
            pass
        # Generate report for salt code coverage
        session.run(
            "coverage",
            "xml",
            "-o",
            str(COVERAGE_REPORT_PROJECT),
            "--omit=tests/*",
            "--include=declarative_requirements/*",
        )
        # Generate report for tests code coverage
        session.run(
            "coverage",
            "xml",
            "-o",
            str(COVERAGE_REPORT_TESTS),
            "--omit=declarative_requirements/*",
            "--include=tests/*",
        )

        session.run(
            "coverage",
            "report",
            "--show-missing",
            "--include=declarative_requirements/*,tests/*",
        )
        # Move the coverage DB to artifacts/coverage in order for it to be archived by CI
        if COVERAGE_REPORT_DB.exists():
            shutil.move(
                str(COVERAGE_REPORT_DB), str(ARTIFACTS_DIR / COVERAGE_REPORT_DB.name)
            )


class Recompress:
    """
    Helper class to re-compress a ``.tag.gz`` file to make it reproducible.
    """

    def __init__(self, mtime):
        self.mtime = int(mtime)

    def tar_reset(self, tarinfo):
        """
        Reset user, group, mtime, and mode to create reproducible tar.
        """
        tarinfo.uid = tarinfo.gid = 0
        tarinfo.uname = tarinfo.gname = "root"
        tarinfo.mtime = self.mtime
        if tarinfo.type == tarfile.DIRTYPE:
            tarinfo.mode = 0o755
        else:
            tarinfo.mode = 0o644
        if tarinfo.pax_headers:
            raise ValueError(tarinfo.name, tarinfo.pax_headers)
        return tarinfo

    def recompress(self, targz):
        """
        Re-compress the passed path.
        """
        tempd = pathlib.Path(tempfile.mkdtemp()).resolve()
        d_src = tempd.joinpath("src")
        d_src.mkdir()
        d_tar = tempd.joinpath(targz.stem)
        d_targz = tempd.joinpath(targz.name)
        with tarfile.open(d_tar, "w|") as wfile:
            with tarfile.open(targz, "r:gz") as rfile:
                rfile.extractall(d_src)
                extracted_dir = next(pathlib.Path(d_src).iterdir())
                for name in sorted(extracted_dir.rglob("*")):
                    wfile.add(
                        str(name),
                        filter=self.tar_reset,
                        recursive=False,
                        arcname=str(name.relative_to(d_src)),
                    )

        with open(d_tar, "rb") as rfh:
            with gzip.GzipFile(
                fileobj=open(d_targz, "wb"), mode="wb", filename="", mtime=self.mtime
            ) as gz:
                while True:
                    chunk = rfh.read(1024)
                    if not chunk:
                        break
                    gz.write(chunk)
        targz.unlink()
        shutil.move(str(d_targz), str(targz))


@nox.session(python="3")
def build(session):
    """
    Build source and binary distributions based off the current commit author
    date UNIX timestamp.

    The reason being, reproducible packages.
    .. code-block: shell
        git show -s --format=%at HEAD
    """
    shutil.rmtree("dist/", ignore_errors=True)
    if SKIP_REQUIREMENTS_INSTALL is False:
        session.install(
            "--progress-bar=off",
            "-r",
            "requirements/build.txt",
            silent=PIP_INSTALL_SILENT,
        )

    timestamp = session.run(
        "git",
        "show",
        "-s",
        "--format=%at",
        "HEAD",
        silent=True,
        log=False,
        stderr=None,
    ).strip()
    env = {"SOURCE_DATE_EPOCH": str(timestamp)}
    session.run(
        "python",
        "-m",
        "build",
        "--sdist",
        "--wheel",
        str(REPO_ROOT),
        env=env,
    )
    # Recreate sdist to be reproducible
    recompress = Recompress(timestamp)
    for targz in REPO_ROOT.joinpath("dist").glob("*.tar.gz"):
        session.log("Re-compressing %s...", targz.relative_to(REPO_ROOT))
        recompress.recompress(targz)

    sha256sum = shutil.which("sha256sum")
    if sha256sum:
        packages = [
            str(pkg.relative_to(REPO_ROOT))
            for pkg in REPO_ROOT.joinpath("dist").iterdir()
        ]
        session.run("sha256sum", *packages, external=True)
    session.run("python", "-m", "twine", "check", "dist/*")
